### PR TITLE
Jenkins default reports

### PR DESCRIPTION
**Pull request by djbowers**
The new default reports menu can be found from the main menu under 
`[j] Jenkins menu` -> `[d] Default Reports`

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+CUSTOM_PARAMS_PATH = 'conf/custom_params.cfg'
+
+JENKINS_URL = ''
+JENKINS_BRANCHES = ''
+JENKINS_PROJECT = ''
+
+custom_params_path = CUSTOM_PARAMS_PATH
+if not os.path.exists(custom_params_path):
+    print('WARNING! Cannot find conf/custom_params.cfg. Will not have default project nor jenkins config data.')
```
Comment (josh-mckenzie) [inline config.py]: From a UX perspective, we should link to a .md doc w/recommendations or an example of what this looks like and what it does.

File: config.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    JENKINS_URL = 'https://test.jenkins.com'
```
Comment (josh-mckenzie) [inline config.py]: Is this a functional URL? Doesn't look like it.

File: src/jenkins_default_reports.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import json
+import os
+import tempfile
+import traceback
+from functools import reduce
+from multiprocessing import Pool
+from os.path import exists
+from urllib.request import urlopen
+
+import config
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: nit: should this import be above w/the others?

File: src/jenkins_default_reports.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+from src.utils import clear, load_file, open_url_in_browser, pause, pick_value
+
+JENKINS_URL = config.JENKINS_URL
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: Why alias instead of import?

File: src/jenkins_default_reports.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+JENKINS_BRANCHES = config.JENKINS_BRANCHES
+JENKINS_PROJECT = config.JENKINS_PROJECT
+
+BUILDS_TO_CHECK = 50
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: Is this configurable via options? And where did our default #'s come from?

File: src/jenkins_default_reports.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+
+def generate_branch_report(report_type):
+    """
+    Generate a report for a branch
+    """
+    branch = pick_value('Which branch would you want to get a test report for? ',
+                        JENKINS_BRANCHES, True, 'Cancel')
+    if not branch:
+        print("Invalid selection, please try again.")
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: If someone cancel's out and branch is null, they'll get a notification that it's an invalid selection. Somewhat crufty from a UX perspective.

File: src/jenkins_default_reports.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        traceback.print_stack()
+        return
+
+    builds = json.loads(response.read())
+    orig_build_numbers = set(list(map(lambda build: build['number'], builds['builds']))[:BUILDS_TO_CHECK])
+
+    pool = Pool(processes=8)
+    print("Downloading builds...")
+
+    if not exists(_get_argus_dir()):
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: What do you think about making _get_*_dir() methods create the dir if they don't exist?

File: src/jenkins_default_reports.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    max_length = max(map(lambda x: len(x[0][0]) + len(x[0][1]), test_cases))
+
+    clear()
+
+    def print_separator():
+        print('-' * (max_length + 5 + 3 + 35))
+
+    print_separator()
+    fmt = '| {:3d} | {:' + str(max_length + 5) + 's} | Failed {:3d} out of {:3d} times |'
```
Comment (josh-mckenzie) [inline src/jenkins_default_reports.py]: Without opening up the code-base to dig into this (have 5 minutes left before next meeting) - does this functionality exist elsewhere? Not just the formatting here, but the general "here's test cases, here's the failure count, etc"? I'd expect some of this logic to already exist in jenkins_report.py and the implementation here to be duplication / i.e. violation of DRY. Sorry if I'm off-base here; not terribly familiar with the jenkins reporting code.

File: src/jenkins_default_reports.py (hunk 3 above)
Comment (ptnapoleon) [reply src/jenkins_default_reports.py]: Reverse nit: It should not. Python imports should go:
```
stdlib

installed dependencies

local imports
```

Comment (josh-mckenzie) [reply src/jenkins_default_reports.py]: Got it. Assumed it was a stdlib due to inconsistency w/semantics (relative path as src.utils vs. blank 'config')

Comment (ptnapoleon) [reply src/jenkins_default_reports.py]: True, maybe this file should be under `src/`?

**Pull request closed without merging**